File: ESCAlertController/ESCAlertController/AlertViewController/ESCAlertViewController.m
//
//  ESCAlertViewController.m
//  ESCAlertController
//
//  Created by xiatian on 2023/6/8.
//


#import "ESCAlertViewController.h"
#import "Masonry.h"


@interface ESCAlertViewController ()

@property(nonatomic,copy)NSString* message;

@property(nonatomic,copy)NSString* titleString;

@property(nonatomic,weak)UIView* contentView;

@property(nonatomic,strong)NSMutableArray<ESCAlertAction *>* actionArray;

@end

@implementation ESCAlertViewController


+ (instancetype)alertControllerWithTitle:(NSString *)title message:(NSString *)message {
    ESCAlertViewController *alertController = [[ESCAlertViewController alloc] init];
    alertController.message = message;
    alertController.titleString = title;
    return alertController;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupUI];
    
}

- (void)addAction:(ESCAlertAction *)action {
    if (action != nil && [action isKindOfClass:[ESCAlertAction class]]) {
        [self.actionArray addObject:action];
    }
}

- (UIModalPresentationStyle)modalPresentationStyle {
    return UIModalPresentationOverFullScreen;
}

- (UIModalTransitionStyle)modalTransitionStyle {
    return UIModalTransitionStyleCrossDissolve;
}

- (void)setupUI {
    //设置内容view
    self.view.backgroundColor = [UIColor colorWithWhite:0 alpha:0.6];
    
    UIView *contentView = [[UIView alloc] init];
    [self.view addSubview:contentView];
    contentView.backgroundColor = [UIColor whiteColor];
    contentView.layer.cornerRadius = 6;
    
    CGFloat contentViewHeight = 32;
    CGFloat titleLabelHeight = 16;
    contentViewHeight += titleLabelHeight;
    contentViewHeight += 12;

    CGFloat contentLabelHeight = 0;
    
    // 调整行间距
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:self.message];
    NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
    // 设置文字居中
    paragraphStyle.alignment = NSTextAlignmentCenter;
    [paragraphStyle setLineSpacing:5];
    [attributedString addAttribute:NSParagraphStyleAttributeName value:paragraphStyle range:NSMakeRange(0, [self.message length])];
    [attributedString  addAttribute:NSForegroundColorAttributeName value:[UIColor blackColor] range:NSMakeRange(0, [self.message length])];
    [attributedString addAttribute:NSFontAttributeName value:[UIFont systemFontOfSize:14] range:NSMakeRange(0, [self.message length])];
    
    if (self.message) {
        CGSize infoSize = CGSizeMake(300 - 64, 100);
//        NSDictionary *dic = @{NSFontAttributeName : FFTextFont(14)};
        //默认的
//        [self.message boundingRectWithSize:infoSize options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingUsesFontLeading attributes:dic context:nil];
        
        CGRect infoRect = [attributedString boundingRectWithSize:infoSize options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingUsesFontLeading  context:nil];
        
        contentLabelHeight = infoRect.size.height + 1;
    }
    contentViewHeight += contentLabelHeight;
    contentViewHeight += 32;
    contentViewHeight += 1;
    CGFloat buttonHeight = 48;
    contentViewHeight += buttonHeight;

    //
    [contentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.view);
        make.centerX.equalTo(self.view);
        make.width.offset(300);
        make.height.offset(contentViewHeight);
    }];
    
    
    UILabel *titleLabel;
    if (self.titleString && self.titleString.length > 0) {
        //设置标题
        //设置内容
        titleLabel = [[UILabel alloc] init];
        [contentView addSubview:titleLabel];
        [titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(contentView).offset(32);
            make.right.equalTo(contentView).offset(-32);
            make.top.equalTo(contentView).offset(32);
            make.height.offset(titleLabelHeight);
        }];
        titleLabel.textAlignment = NSTextAlignmentCenter;
        titleLabel.textColor = [UIColor blackColor];
        titleLabel.font = [UIFont boldSystemFontOfSize:16];
        titleLabel.text = self.titleString;
    }
    
    //设置内容
    UILabel *contentLabel = [[UILabel alloc] init];
    [contentView addSubview:contentLabel];
    if (titleLabel) {
        [contentLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.centerX.equalTo(contentView);
            make.height.offset(contentLabelHeight);
            make.top.equalTo(titleLabel.mas_bottom).offset(13);
            make.width.lessThanOrEqualTo(contentView).offset(-64);
        }];
    }

    contentLabel.textAlignment = NSTextAlignmentCenter;
    contentLabel.numberOfLines = 3;
    contentLabel.textColor = [UIColor blackColor];
    contentLabel.font = [UIFont systemFontOfSize:14];
    
    
    contentLabel.attributedText = attributedString;
    
    UIView *spaceView = [[UIView alloc] init];
    [contentView addSubview:spaceView];
    spaceView.backgroundColor = [UIColor grayColor];
    [spaceView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(contentView);
        make.top.equalTo(contentLabel.mas_bottom).offset(32);
        make.height.offset(1);
    }];
    
    //设置按钮
    int actionCount = (int)self.actionArray.count;
    NSMutableArray *buttonArray = [NSMutableArray array];
    for (int i = 0; i < actionCount; i++) {
        ESCAlertAction *action = [self.actionArray objectAtIndex:i];
        UIButton *actionButton = [[UIButton alloc] init];
        [actionButton setTitle:action.title forState:UIControlStateNormal];
        [actionButton setTitleColor:action.titleColor forState:UIControlStateNormal];
        actionButton.titleLabel.font = [UIFont systemFontOfSize:15];
        [buttonArray addObject:actionButton];
        [contentView addSubview:actionButton];
        [actionButton addTarget:self action:@selector(didClickButton:) forControlEvents:UIControlEventTouchUpInside];

        actionButton.tag = i;
    }
    
    
    if (actionCount == 0) {
        
    }else if (actionCount == 1) {
        UIButton *button = [buttonArray firstObject];
        [button mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.right.bottom.equalTo(contentView);
            make.top.equalTo(spaceView.mas_bottom);
        }];
    }else if (actionCount == 2) {
        UIView *centerView = [[UIView alloc] init];
        [contentView addSubview:centerView];
        centerView.backgroundColor = [UIColor grayColor];
        [centerView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.centerX.equalTo(contentView);
            make.bottom.equalTo(contentView);
            make.width.offset(1);
            make.top.equalTo(spaceView.mas_bottom);
        }];
        
        UIButton *firstButton = [buttonArray firstObject];
        [firstButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.bottom.equalTo(contentView);
            make.top.equalTo(spaceView.mas_bottom);
            make.right.equalTo(centerView.mas_left);
        }];
        UIButton *lastButton = [buttonArray lastObject];
        [lastButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(centerView.mas_right);
            make.top.equalTo(spaceView.mas_bottom);
            make.bottom.right.equalTo(contentView);
        }];
    }
}

- (void)didClickButton:(UIButton *)button {
    if(self.isFirstDismissViewController == YES){
        int index = (int)button.tag;
        self.view.userInteractionEnabled = NO;
        ESCAlertAction *action = [self.actionArray objectAtIndex:index];
        [self dismissViewControllerAnimated:YES completion:^{
            action.handler(action);
        }];
    }else {
        int index = (int)button.tag;
        self.view.userInteractionEnabled = NO;
        ESCAlertAction *action = [self.actionArray objectAtIndex:index];
        action.handler(action);
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self dismissViewControllerAnimated:YES completion:nil];
        });
    }
}

- (NSMutableArray *)actionArray {
    if (_actionArray == nil) {
        _actionArray = [NSMutableArray array];
    }
    return _actionArray;
}


@end

File: ESCAlertController/Swift_ESCAlertController/AlertViewController/ESCAlertViewController.swift
//
//  ESCAlertViewController.swift
//  Swift_ESCAlertController
//
//  Created by xiatian on 2023/6/8.
//

import UIKit

class ESCAlertViewController: UIViewController {
    
    var isFirstDismissViewController:Bool = false
    
    var message:String?
    
    var titleString:String?

    var contentView:UIView?

    var actionArray:[ESCAlertAction]?
    
  

    init(isFirstDismissViewController: Bool, message: String? = nil, titleString: String? = nil, contentView: UIView? = nil, actionArray: [ESCAlertAction]? = nil) {
        
        self.isFirstDismissViewController = isFirstDismissViewController
        self.message = message
        self.titleString = titleString
        self.contentView = contentView
        self.actionArray = actionArray
        super.init(nibName: nil, bundle: nil)
        self.modalPresentationStyle = .overFullScreen
        self.modalTransitionStyle = .crossDissolve
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    
    @objc func didClickButton(button:UIButton) {
        if self.isFirstDismissViewController == true {
            let index = button.tag
            self.view.isUserInteractionEnabled = false
            let action = self.actionArray![index]
            self.dismiss(animated: true) {
                action.handler!(action)
            }
        }else{
            let index = button.tag
            self.view.isUserInteractionEnabled = false
            let action = self.actionArray![index]
            action.handler!(action)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3, execute: DispatchWorkItem(block: {
                self.dismiss(animated: true)
            }))

        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        setupUI()
    }
    
    func setupUI() {
        //设置内容view
        self.view.backgroundColor = .init(white: 0, alpha: 0.6)
        
        let contentView = UIView()
        self.view.addSubview(contentView)
        contentView.backgroundColor = .white
        contentView.layer.cornerRadius = 6

        var contentViewHeight = 32
        var titleLabelHeight = 16
        contentViewHeight += titleLabelHeight
        contentViewHeight += 12

        var contentLabelHeight = 0;
        
        let attributedString = NSMutableAttributedString(string: self.message!)
        let paragraphStyle = NSMutableParagraphStyle()
        // 设置文字居中
        paragraphStyle.alignment = NSTextAlignment.center
        paragraphStyle.lineSpacing = 5
        attributedString.addAttribute(NSAttributedString.Key.paragraphStyle, value: paragraphStyle, range: NSRange(location: 0, length: self.message!.count))
        attributedString.addAttribute(NSAttributedString.Key.foregroundColor, value: UIColor.black, range: NSRange(location: 0, length: self.message!.count))
        attributedString.addAttribute(NSAttributedString.Key.font, value: UIFont.systemFont(ofSize: 14), range: NSRange(location: 0, length: self.message!.count))

        if let message = self.message {
            let infoSize = CGSize(width: 300 - 64, height: 100)
            // 默认的
            // let dic = [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 14)]
            // message.boundingRect(with: infoSize, options: [.usesLineFragmentOrigin, .usesFontLeading], attributes: dic, context: nil)
            
            let infoRect = attributedString.boundingRect(with: infoSize, options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
            
            contentLabelHeight = Int(infoRect.size.height + 1)
        }
        
        
        
        contentViewHeight += contentLabelHeight;
        contentViewHeight += 32;
        contentViewHeight += 1;
        var buttonHeight = 48;
        contentViewHeight += buttonHeight;
        
        contentView.snp.makeConstraints { make in
            make.centerY.equalToSuperview()
            make.centerX.equalToSuperview()
            make.width.equalTo(300)
            make.height.equalTo(contentViewHeight)
        }
        
      
        
        var titleLabel: UILabel?
        if let titleString = self.titleString, !titleString.isEmpty {
            // 设置标题
            titleLabel = UILabel()
            contentView.addSubview(titleLabel!)
            titleLabel!.textAlignment = .center
            titleLabel!.textColor = .black
            titleLabel!.font = UIFont.boldSystemFont(ofSize: 16)
            titleLabel!.text = titleString
            titleLabel!.snp.makeConstraints { make in
                make.left.equalTo(contentView).offset(32)
                make.right.equalTo(contentView).offset(-32)
                make.top.equalTo(contentView).offset(32)
                make.height.equalTo(titleLabelHeight)
            }
        }
 
        // 设置内容
        let contentLabel = UILabel()
        contentView.addSubview(contentLabel)
        contentLabel.textAlignment = .center
        contentLabel.numberOfLines = 3
        contentLabel.textColor = .black
        contentLabel.font = UIFont.systemFont(ofSize: 14)
        contentLabel.attributedText = attributedString
        contentLabel.snp.makeConstraints { make in
            make.centerX.equalTo(contentView)
            make.height.equalTo(contentLabelHeight)
            if let titleLabel = titleLabel {
                make.top.equalTo(titleLabel.snp.bottom).offset(13)
                make.width.lessThanOrEqualTo(contentView).offset(-64)
            } else {
                make.top.equalTo(contentView).offset(32)
                make.left.right.equalTo(contentView).inset(32)
            }
        }

        let spaceView = UIView()
        contentView.addSubview(spaceView)
        spaceView.backgroundColor = .gray
        spaceView.snp.makeConstraints { make in
            make.left.right.equalTo(contentView)
            make.top.equalTo(contentLabel.snp.bottom).offset(32)
            make.height.equalTo(1)
        }
       
        
        
       
        var buttonArray = [UIButton]()
        let actionCount = self.actionArray!.count
        for i in 0..<actionCount {
            let action = self.actionArray![i]
            let actionButton = UIButton()
            actionButton.setTitle(action.title, for: .normal)
            actionButton.setTitleColor(action.titleColor, for: .normal)
            actionButton.titleLabel?.font = UIFont.systemFont(ofSize: 15)
            buttonArray.append(actionButton)
            contentView.addSubview(actionButton)
            actionButton.addTarget(self, action: #selector(didClickButton(button:)), for: .touchUpInside)
            actionButton.tag = i
        }

        if actionCount == 0 {
            // Do nothing
        } else if actionCount == 1 {
            let button = buttonArray.first!
            button.snp.makeConstraints { make in
                make.left.right.bottom.equalTo(contentView)
                make.top.equalTo(spaceView.snp.bottom)
            }
        } else if actionCount == 2 {
            let centerView = UIView()
            contentView.addSubview(centerView)
            centerView.backgroundColor = .gray
            centerView.snp.makeConstraints { make in
                make.centerX.equalTo(contentView)
                make.bottom.equalTo(contentView)
                make.width.equalTo(1)
                make.top.equalTo(spaceView.snp.bottom)
            }

            let firstButton = buttonArray.first!
            firstButton.snp.makeConstraints { make in
                make.left.bottom.equalTo(contentView)
                make.top.equalTo(spaceView.snp.bottom)
                make.right.equalTo(centerView.snp.left)
            }

            let lastButton = buttonArray.last!
            lastButton.snp.makeConstraints { make in
                make.left.equalTo(centerView.snp.right)
                make.top.equalTo(spaceView.snp.bottom)
                make.bottom.right.equalTo(contentView)
            }
        }
       
    }
    
    public func addAction(action:ESCAlertAction) {
        if self.actionArray == nil {
            self.actionArray = [action]
        }else {
            self.actionArray?.append(action)
        }
    }

}
